subprojects/language-native/src/integTest/groovy/org/gradle/language/cpp/CppCachingIntegrationTest.groovy

```diff
@@ -50,30 +53,99 @@ class CppCachingIntegrationTest extends AbstractCppInstalledToolChainIntegration
         }
     }
 
-    def 'compilation can be cached'() {
-        setupProjectInDirectory(temporaryFolder.testDirectory)
+    @Unroll
+    def 'compilation can be cached (#buildType)'() {
+        setupProject()
+        this.buildType = buildType
 
         when:
-        withBuildCache().succeeds compilationTask
+        withBuildCache().run compilationTask
 
         then:
         compileIsNotCached()
 
         when:
-        withBuildCache().succeeds 'clean', ':installDebug'
+        withBuildCache().run 'clean', installTask(buildType)
 
         then:
         compileIsCached()
-        installation("build/install/main/debug").exec().out == app.expectedOutput
+        installation("build/install/main/${buildType.toLowerCase()}").exec().out == app.expectedOutput
+
+        where:
+        buildType << [debug, release]
+    }
+
+    def "compilation task is relocatable"() {
+        buildType = release
+        def originalLocation = file('original-location')
+        def newLocation = file('new-location')
+        setupProject(originalLocation)
+        setupProject(newLocation)
+
+        when:
+        inDirectory(originalLocation)
+        withBuildCache().run compilationTask
+
+        def snapshotsInOriginalLocation = snapshotObjects(originalLocation)
+
+        then:
+        executedAndNotSkipped compilationTask
+
+        when:
+        executer.beforeExecute {
+            usingProjectDirectory(newLocation)
+        }
+        run compilationTask
+
+        then:
+        executedAndNotSkipped(compilationTask)
+        assertSameSnapshots(snapshotsInOriginalLocation, snapshotObjects(newLocation))
+
+        when:
+        run 'clean'
+        withBuildCache().run compilationTask, installTask(buildType)
+
+        then:
+        skipped compilationTask
+        assertSameSnapshots(snapshotsInOriginalLocation, snapshotObjects(newLocation))
+        installation(newLocation.file("build/install/main/${buildType.toLowerCase()}")).exec().out == app.expectedOutput
+    }
+
+    private void assertSameSnapshots(Map<String, TestFile.Snapshot> snapshotsInOriginalLocation, Map<String, TestFile.Snapshot> snapshotsInNewLocation) {
+        assert snapshotsInOriginalLocation.keySet() == snapshotsInNewLocation.keySet()
+        if (nonDeterministicCompilation || (toolChain.displayName == "clang" && OperatingSystem.current().isLinux())) {
```

Comment:
Why is Clang on Linux singled out here?

Reply:
It embeds absolute paths into the object files. Let's see if we can fix it with my latest commit.

Reply:
OK, but even if that's the case, shouldn't it be determined by `isNonDeterministicCompilation()`?

Reply:
No, since the compilation is deterministic - it just is not relocatable.

Reply:
Should we call that out in a similar method, like `isNonRelocatableCompilation()`?

Reply:
I extracted the method.